### ISSUE TITLE
rlvr: build the GRPO trainer that all four training runs share

### ISSUE THREAD

**Issue by asuzukosi** [asuzukosi/ai-experiments#9]
**Code only — no full-scale run happens here.** One module and four YAML configs; every run differs by configuration, not by code. Splitting the runs out means the dry run can sit between this and them, and means the spec's cut order (drop the unfiltered control first, then the no-KL ablation, never the random-reward control) is expressible as separate issues rather than a note.

**Rollout budget is a first-class hyperparameter.** Published results at fixed step count: 43.2% (256 tokens, 4 rollouts) -> 45.6% (512, 4) -> 47.4% (512, 8). A 4-point swing from generation budget alone, larger than most ablations here. Record the cap with every result — and response length *grows* during training, so a tight cap tightens further.

### Sketch

```python
# src/rlvr/grpo.py — one module, four configs
def make_reward(arm: str):
    if arm == "random":
        return lambda completions, **kw: [float(random.random() < 0.5) for _ in completions]
    return lambda completions, **kw: [total_reward(c, g) for c, g in zip(completions, kw["gold"])]

def build_trainer(arm: str, model="Qwen/Qwen2.5-1.5B"):
    cfg = GRPOConfig(
        output_dir=f"results/checkpoints/{run_name('qwen2.5-1.5b', f'grpo-{arm}', now)}",
        num_generations=8,                 # 8-16; RECORD IT
        max_completion_length=768,         # RECORD IT — the 4-point swing lives here
        learning_rate=1e-6, epsilon=0.2,
        beta=0.0 if arm == "nokl" else 0.01,
        use_vllm=True, vllm_mode="colocate",
        save_strategy="steps", save_steps=25, save_total_limit=3,
        max_steps=200 if arm == "random" else 500,
        report_to="wandb",
    )
    return GRPOTrainer(model=model, reward_funcs=[make_reward(arm)],
                       args=cfg, train_dataset=ds_for(arm))
```

```
configs/
├── grpo.yaml            filtered prompts, verifier + format reward, beta 0.01
├── grpo_unfiltered.yaml random prompts, equal size, equal compute
├── grpo_nokl.yaml       beta 0
└── grpo_random.yaml     reward = fair coin, ~200 steps
```

### TODO
- [ ] `scripts/train_grpo.py --arm {main,unfiltered,nokl,random}` — arms differ only by YAML
- [ ] 8-16 generations/prompt, lr 1e-6, clip 0.2, max completion 768
- [ ] Penalise or filter overlong generations; score only EOS-terminated completions
- [ ] Checkpoint every 25-50 steps to the volume so a killed node loses nothing
- [ ] Verify resume restores policy + optimizer + step, and W&B continues on the same run id
- [ ] **Smoke-test on a few dozen prompts** — the full runs are separate issues

### References
- TRL GRPOTrainer: https://huggingface.co/docs/trl/main/en/grpo_trainer
- GRPO / group-relative advantage, DeepSeek-R1: https://arxiv.org/abs/2501.12948
- TRL vLLM integration: https://huggingface.co/docs/trl/main/en/speeding_up_training

### Compute

Smoke test only — minutes on any GPU. The runs are #39-#42.

---

### Dependencies
- **Blocked by:** #6, #8
- **Blocks:** #10, #15, #35, #39, #40, #41, #42

Self-contained within this project — no cross-project dependencies.
